integration/emr/alluxio-emr.sh

```diff
@@ -10,99 +10,217 @@
 #
 
-# This script is meant for bootstrapping the Alluxio service to an EMR cluster. Arguments for the script are listed below.
+# This script is meant for bootstrapping the Alluxio service to an EMR cluster.
+# Arguments for the script are listed below.
 # Arg 1. Download URI (ex. http://downloads.alluxio.io/downloads/files/2.0.0/alluxio-2.0.0-bin.tar.gz)
 # Arg 2. Root UFS URI (ex. s3://my-bucket/alluxio-emr/mount)
-# Arg 3. Extra Alluxio Options. These will be appended to alluxio-site.properties. Multiple options can be specified using ';' as a delimiter
-#        (ex. alluxio.user.file.writetype.default=CACHE_THROUGH;alluxio.user.file.readtype.default=CACHE)
-
-#Create user
-sudo groupadd alluxio -g 600
-sudo useradd alluxio -u 600 -g 600
-
-#Download the release
-#TODO Add metadata header tag to the wget for filtering out in download metrics.
-if [[ -z $1 ]]
-then
-  echo "No Download URL Provided. Please go to http://downloads.alluxio.io to see available release downloads."
-else
-  if [[ "${1}" == s3://* ]]
+# Arg 3. (Optional) HTTP(S) or S3 URIs pointing to alluxio-site.properties file
+#        that will be used on cluster startup. The resource must be named "alluxio-site.properties"
+# Arg 4. (Optional) Delimeter for additional properties
+# Arg 5. (Optional) Extra Alluxio Options. These will be appended to
+#        alluxio-site.properties. Multiple options can be specified using Arg 4
+#        as a delimiter (ex. alluxio.user.file.writetype.default=CACHE_THROUGH;alluxio.user.file.readtype.default=CACHE)
+
+set -o errexit  # exit when a command fails - append "|| true" to allow a
+                # command to fail
+set -o nounset  # exit when attempting to use undeclared variables
+
+
+set -x # debugging
+
+# script constants
+ALLUXIO_HOME=/opt/alluxio
+ALLUXIO_SITE_PROPERTIES=${ALLUXIO_HOME}/conf/alluxio-site.properties
+
+# Downloads a file to the local machine from a remote HTTP(S) or S3 URI into the cwd
+#
+# Args:
+#   $1: URI - the remote location to retrieve the file from
+download_file() {
+  local uri=$1
+
+  if [[ "${uri}" == s3://* ]]
+  then
+    aws s3 cp ${uri} ./
+  else
+    # TODO Add metadata header tag to the wget for filtering out in download metrics.
+    wget -nv "${uri}"
+  fi
+}
+
+# Run a command as a specific user
+# Assumes the provided user already exists on the system and user running script has sudo access
+#
+# Args:
+#   $1: user - the username to run the command as
+#   $2: command - the command to run
+doas() {
+  local user=$1
+  local cmd=$2
+  sudo runuser -l ${user} -c "${cmd}"
+}
+
+# Appends a property KV pair to the alluxio-site.properties file
+#
+# Args:
+#   $1: property name
+#   $2: property value
+append_alluxio_property() {
+  local property=$1
+  local value=$2
+
+  # OK to fail in this section
+  set +o errexit
+  # /opt/alluxio/conf must exist for this to work
+  grep -q ${property} ${ALLUXIO_SITE_PROPERTIES} 2> /dev/null
```

Comment:
What happens if a property is a substring of an existing property? Can we do an exact string match? With some of the templates, this can get a little dicey.

Ex:
alluxio.master.mount.table.root.option.alluxio.underfs.version=2.7
alluxio.underfs.version=3.1

Reply:
Ooh, good catch. I think the quick fix for this is for grep to use `${property}=`

Reply:
I can also use regex to force matching from beginning of the line

    grep -q -e "^${property}="

Reply:
I think our current file allows whitespace at the beginning of the line. So if we expect users to maybe try out a local mode alluxio-site.properties in EMR, then we should have consistent behavior.

Reply:
Regex for allowing the beginning whitespace is `^\s*${property}=`